### PR TITLE
Sample script:  Airship

### PR DESCRIPTION
**Pull request by PhilipRosedale**
Example of script (makeAirship.js) that makes a floating ship that slowly follows people in your domain and shoots cannonballs at them that make loud explosion sounds and lights.  

Change lifetime in makeAirship.js, right now it is set to 2 days.  

Good example of an entity script that can be left and will activate when people are nearby.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7e658896e63a1b7e8728a719d41e5890a46a2397 available at: https://gist.github.com/d2d74bdd05654cbd8d96<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7438.dmg

File: examples/airship/airship.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        timeoutID = undefined,
+        bullet = null,
+        particles = null,
+        nearbyAvatars = 0,
+        nearbyAvatarsInRange = 0,
+        averageAvatarLocation = { x: 0, y: 0, z: 0 },
+        properties,
+        lightTimer = 0,
+        lightTimeoutID = undefined,
+        audioInjector = null
```
Comment (ZappoMan) [inline examples/airship/airship.js]: missing `;`

File: examples/airship/airship.js [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+            lifetime: timeOfFlight * 2
+        });
+
+        Audio.playSound(cannonSound, {
+            position: Vec3.sum(MyAvatar.position, velocity),
+            volume: 1.0
+        });
+        makeParticles(properties.position, bullet, timeOfFlight * 2);
+        Script.setTimeout(explode, timeOfFlight * 1000);
+
```
Comment (ZappoMan) [inline examples/airship/airship.js]: cleanup some of this white space

File: examples/airship/airship.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        if (nearbyAvatars && (Math.random() < 1 / nearbyAvatars)) {
+            move();
+        }
+
+        if (nearbyAvatarsInRange && (Math.random() < PROBABILITY_OF_SHOOT / nearbyAvatarsInRange)) {
+            shoot();
+        }
+
+        var timeToNextCheck = 33;
```
Comment (ZappoMan) [inline examples/airship/airship.js]: for consistency... `var TIME_TO_NEXT_CHECK = 33;`

File: examples/airship/airship.js [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+            Script.clearTimeout(lightTimeoutID);
+        }
+        if (spotlight !== null) {
+            Entities.deleteEntity(spotlight);
+        }
+    };
+
+    function makeParticles(position, parent, lifespan) {
+        particles = Entities.addEntity({
+        type: 'ParticleEffect',
```
Comment (ZappoMan) [inline examples/airship/airship.js]: lines 255 to 301 should be indented 4 spaces

**Comment by OmegaHeron** [comment]
I admit I haven't played with entity scripts much yet - so I may need some education.  Should I place airship.js in a entity via its script url parameter?  Did that (modified to pull makeairship.js from my http server) and I got shot at, booms - cannon ball and light, but never saw ship.  Server spit out a bunch of;

[03/23 14:13:49] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{d67b27b8-88cf-4bb8-865a-ea089fe48004}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:13:49] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{d67b27b8-88cf-4bb8-865a-ea089fe48004}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:13:49] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{d67b27b8-88cf-4bb8-865a-ea089fe48004}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:00] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{cb125668-7be3-4659-a9f9-b6f42ca9628f}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:00] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{cb125668-7be3-4659-a9f9-b6f42ca9628f}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:00] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{cb125668-7be3-4659-a9f9-b6f42ca9628f}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:00] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{cb125668-7be3-4659-a9f9-b6f42ca9628f}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:00] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{cb125668-7be3-4659-a9f9-b6f42ca9628f}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:01] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{c7a0ee9a-44f9-46cd-9b61-ced36cd51958}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:01] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{c7a0ee9a-44f9-46cd-9b61-ced36cd51958}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:03] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{511c20c4-dd84-4e39-ab84-22b2de01ab97}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:03] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{511c20c4-dd84-4e39-ab84-22b2de01ab97}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:03] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{511c20c4-dd84-4e39-ab84-22b2de01ab97}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:03] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{511c20c4-dd84-4e39-ab84-22b2de01ab97}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )
[03/23 14:14:03] [DEBUG] Recieved packet for previously deleted entity [ QUuid("{511c20c4-dd84-4e39-ab84-22b2de01ab97}") ] ignoring. (inside  EntityTreeElement::readElementDataFromBuffer )

During run.

**Comment by ZappoMan** [comment]
@OmegaHeron - those server logs are benign log spam... I will file a P2 bug to remove it... I had meant to do that before.

**Comment by imgntn** [comment]
@OmegaHeron i think u can run this spawner https://cdn.rawgit.com/PhilipRosedale/hifi/airship/examples/airship/makeAirship.js

**Comment by OmegaHeron** [comment]
Thanks @imgntn - I'll give that a try later.  I had picked up scripts and modified calling URL to match my setup - but never got an airship - just bwhoum balls.  I just wanted to see it work ;)  Not done a lot of entity scripting as I mentioned to start so - figure I'm doing something wrong.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7b326c9d7bdf8daafb3168ce1fbadc4af14f5999 available at: https://gist.github.com/d2d74bdd05654cbd8d96<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7438.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8c4ecc64d94f6a5a99994d00638ec89b080ada04 available at: https://gist.github.com/d2d74bdd05654cbd8d96<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7438.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7438.dmg

**Comment by PhilipRosedale** [comment]
Test plan:  

Run the spawner script, and what should happen is that you get a floating ship showing up in front of you and above you.  It should have rotating propellers.  It should turn and slowly follow the average location of all avatars in the domain.   Once in a while you should see a light on deck, hear a cannon sound, and a cannonball should shoot in the direction on an avatar, and then blow up when it gets to them with another sound and light.  

You need to have edit rights in the domain to see those cannonballs.  

Test whether you can start the ship on one machine, and then have someone new come in from another machine.  The ship should follow and shoot cannonballs at them too.

**Pull request merged**